ghost/link-replacement/lib/link-replacement.js

```diff
@@ -1,33 +1,51 @@
+/**
+ * @typedef {object} ILinkRedirect
+ * @prop {URL} to
+ * @prop {from} to
+ * @prop {from} to
+ */
+
+/**
+ * @typedef {object} ILinkRedirectService
+ * @prop {(to: URL) => Promise<ILinkRedirect>} addRedirect
+ */
+
+/**
+ * @typedef {object} ILinkClickTrackingService
+ * @prop {(link: ILinkRedirect) => Promise<URL>} addTrackingToRedirect
+ */
+
 class LinkReplacementService {
-    /// Placeholder method
-    async createRedirect(url) {
-        return Promise.resolve({
-            from: 'https://example.com/r/' + Math.random().toString(36).substring(2, 15),
-            to: url
-        });
-    }
+    /** @type ILinkRedirectService */
+    #linkRedirectService;
+    /** @type ILinkClickTrackingService */
+    #linkClickTrackingService;
 
-    // Todo: move to different service
-    addTrackingToRedirect(redirect, memberUuid) {
-        const newUrl = new URL(redirect.from);
-        newUrl.searchParams.append('m', memberUuid);
-        return newUrl;
+    /**
+     * @param {object} deps
+     * @param {ILinkRedirectService} deps.linkRedirectService
+     * @param {ILinkClickTrackingService} deps.linkClickTrackingService
+     */
+    constructor(deps) {
+        this.#linkRedirectService = deps.linkRedirectService;
+        this.#linkClickTrackingService = deps.linkClickTrackingService;
     }
 
     async replaceLink(url, newsletter, post) {
         // Can probably happen in one call to the MemberAttributionService (but just to make clear what happens here)
-        
+
         // 1. Add attribution
         // TODO: this should move the the attribution service in the future
         url.searchParams.append('rel', newsletter.get('slug') + '-newsletter');
         url.searchParams.append('attribution_id', post.id);
         url.searchParams.append('attribution_type', 'post');
 
         // 2. Add redirect for link click tracking
-        const redirect = await this.createRedirect(url.toString());
+        const redirect = await this.#linkRedirectService.addRedirect(url);
```

Comment:
If we add `from` as a parameter, what do you think if we would do something like this instead (doesn't have to be in this PR):
```
const redirect = await this.#linkClickTrackingService.createTrackedRedirect(url);
```
Which internally would call `await this.#linkRedirectService.addRedirect(generatedSlug, url);`

Reply:
Maybe we can also pass the member uuid so that it includes that in the returned `from`

Reply:
Essentially we could also name it `this.#linkClickTrackingService.createTrackedLink(url);` which returns a normal URL and hides the internal implementation detail that it works via redirects 🤷‍♂️